// @nutui/nutui-react/dist/es/packages/avatargroup/avatargroup.js
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import React, { useRef } from "react";
import classNames from "classnames";
import { AvatarContext } from "./context";
import { ComponentDefaults } from "../../utils/typings";
var defaultProps = _object_spread_props(_object_spread({}, ComponentDefaults), {
    maxContent: '',
    max: '',
    maxBackground: '#eee',
    maxColor: '#666',
    gap: '-8',
    level: 'left'
});
var classPrefix = "nut-avatar-group";
export var AvatarGroup = function(props) {
    var propAvatarGroup = _object_spread({}, defaultProps, props);
    var className = propAvatarGroup.className, style = propAvatarGroup.style, children = propAvatarGroup.children;
    var avatarGroupRef = useRef(null);
    var cls = classNames(classPrefix, className);
    var parentAvatar = {
        propAvatarGroup: propAvatarGroup,
        avatarGroupRef: avatarGroupRef
    };
    return /*#__PURE__*/ React.createElement(AvatarContext.Provider, {
        value: parentAvatar
    }, /*#__PURE__*/ React.createElement("div", {
        className: cls,
        style: style,
        ref: avatarGroupRef
    }, children));
};
AvatarGroup.displayName = 'NutAvatarGroup';
