// @nutui/nutui-react/dist/es/packages/animatingnumbers/countup.js
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { _ as _to_consumable_array } from "@swc/helpers/_/_to_consumable_array";
import React, { useCallback, useEffect, useMemo, useRef } from "react";
import { ComponentDefaults } from "../../utils/typings";
import { mergeProps } from "../../utils/merge-props";
var defaultProps = _object_spread_props(_object_spread({}, ComponentDefaults), {
    length: 0,
    value: '',
    delay: 300,
    duration: 1,
    thousands: false
});
export var CountUp = function(props) {
    var _mergeProps = mergeProps(defaultProps, props), length = _mergeProps.length, value = _mergeProps.value, delay = _mergeProps.delay, duration = _mergeProps.duration, className = _mergeProps.className, thousands = _mergeProps.thousands, style = _mergeProps.style, rest = _object_without_properties(_mergeProps, [
        "length",
        "value",
        "delay",
        "duration",
        "className",
        "thousands",
        "style"
    ]);
    var classPrefix = 'nut-countup';
    var countupRef = useRef(null);
    var timerRef = useRef(0);
    var numbers = Array.from({
        length: 10
    }, function(v, i) {
        return i;
    });
    var getShowNumber = function() {
        var splitArr = value.split('.');
        var intNumber = length && splitArr[0].length < length ? (Array(length).join('0') + splitArr[0]).slice(-length) : splitArr[0];
        var currNumber = "".concat(thousands ? intNumber.replace(/(\d)(?=(?:\d{3})+$)/g, '$1,') : intNumber).concat(splitArr[1] ? '.' : '').concat(splitArr[1] || '');
        return currNumber.split('');
    };
    var numberArr = useMemo(getShowNumber, [
        value,
        length,
        thousands
    ]);
    var setNumberTransform = useCallback(function() {
        if (countupRef.current) {
            var numberItems = countupRef.current.querySelectorAll('.nut-countup-number');
            var numberFilterArr = numberArr.filter(function(item) {
                return !Number.isNaN(Number(item));
            });
            Object.keys(numberItems).forEach(function(key) {
                var elem = numberItems[Number(key)];
                var idx = Number(numberFilterArr[Number(key)]);
                if ((idx || idx === 0) && elem) {
                    // 计算规则：父元素和实际列表高度的百分比，分割成20等份
                    var transform = "translate(0, -".concat((idx === 0 ? 10 : idx) * 5, "%)");
                    elem.style.transform = transform;
                    elem.style.webkitTransform = transform;
                }
            });
        }
    }, [
        numberArr
    ]);
    var numberEaseStyle = {
        transitionDuration: "".concat(duration, "s")
    };
    useEffect(function() {
        timerRef.current = window.setTimeout(function() {
            setNumberTransform();
        }, delay);
        return function() {
            window.clearTimeout(timerRef.current);
        };
    }, [
        numberArr,
        delay,
        setNumberTransform
    ]);
    return /*#__PURE__*/ React.createElement("div", {
        className: "".concat(classPrefix, " ").concat(className),
        ref: countupRef
    }, /*#__PURE__*/ React.createElement("ul", {
        className: "".concat(classPrefix, "-list")
    }, numberArr.map(function(item, idx) {
        return /*#__PURE__*/ React.createElement("li", {
            className: "".concat(classPrefix, "-listitem ").concat(!Number.isNaN(Number(item)) ? "".concat(classPrefix, "-listitem-number") : ''),
            key: idx
        }, !Number.isNaN(Number(item)) ? /*#__PURE__*/ React.createElement("span", {
            className: "".concat(classPrefix, "-number"),
            style: numberEaseStyle
        }, _to_consumable_array(numbers).concat(_to_consumable_array(numbers)).map(function(number, subidx) {
            return /*#__PURE__*/ React.createElement("span", {
                className: "".concat(classPrefix, "-number-text"),
                key: subidx
            }, number);
        })) : /*#__PURE__*/ React.createElement("span", {
            className: "".concat(classPrefix, "-separator")
        }, item));
    })));
};
CountUp.displayName = 'NutCountUp';
