// @nutui/nutui-react/dist/es/packages/calendar/utils.js
import { Utils } from "../../utils/date";
export var splitDate = function(date) {
    var split = date.indexOf('-') !== -1 ? '-' : '/';
    return date.split(split);
};
export var isMultiple = function(day, days) {
    if (days.length > 0) {
        return days.some(function(item) {
            return Utils.isEqual(item, day);
        });
    }
    return false;
};
export var isCurrDay = function(month, day) {
    var date = "".concat(month.curData[0], "/").concat(month.curData[1], "/").concat(day);
    return Utils.isEqual(date, Utils.date2Str(new Date(), '/'));
};
export var getCurrDate = function(day, month) {
    return "".concat(month.curData[0], "/").concat(month.curData[1], "/").concat(Utils.getNumTwoBit(+day.day));
};
export var isStart = function(day, days) {
    return Utils.isEqual(days[0], day);
};
export var isEnd = function(day, days) {
    return Utils.isEqual(days[1], day);
};
// 开始结束时间是否相等
export var isStartAndEnd = function(days) {
    return days.length >= 2 && Utils.isEqual(days[0], days[1]);
};
