// @nutui/nutui-react/dist/es/utils/get-rect.js
export var inBrowser = typeof document !== 'undefined' && !!document.scripts;
export function isWindow(val) {
    return val === window;
}
/**
  获取元素的大小及其相对于视口的位置，等价于 Element.getBoundingClientRect。
  width 宽度	number
  height 高度	number
  top	顶部与视图窗口左上角的距离	number
  left	左侧与视图窗口左上角的距离	number
  right	右侧与视图窗口左上角的距离	number
  bottom	底部与视图窗口左上角的距离	number
 */ export var getRect = function(elementRef) {
    var element = elementRef;
    if (isWindow(element)) {
        var width = element.innerWidth;
        var height = element.innerHeight;
        return {
            top: 0,
            left: 0,
            right: width,
            bottom: height,
            width: width,
            height: height
        };
    }
    if (element && element.getBoundingClientRect) {
        // console.log(element, element.getBoundingClientRect())
        return element.getBoundingClientRect();
    }
    return {
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        width: 0,
        height: 0
    };
};
