// @nutui/nutui-react/dist/cjs/packages/calendar/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    getCurrDate: function() {
        return getCurrDate;
    },
    isCurrDay: function() {
        return isCurrDay;
    },
    isEnd: function() {
        return isEnd;
    },
    isMultiple: function() {
        return isMultiple;
    },
    isStart: function() {
        return isStart;
    },
    isStartAndEnd: function() {
        return isStartAndEnd;
    },
    splitDate: function() {
        return splitDate;
    }
});
var _date = require("../../utils/date");
var splitDate = function splitDate(date) {
    var split = date.indexOf('-') !== -1 ? '-' : '/';
    return date.split(split);
};
var isMultiple = function isMultiple(day, days) {
    if (days.length > 0) {
        return days.some(function(item) {
            return _date.Utils.isEqual(item, day);
        });
    }
    return false;
};
var isCurrDay = function isCurrDay(month, day) {
    var date = "".concat(month.curData[0], "/").concat(month.curData[1], "/").concat(day);
    return _date.Utils.isEqual(date, _date.Utils.date2Str(new Date(), '/'));
};
var getCurrDate = function getCurrDate(day, month) {
    return "".concat(month.curData[0], "/").concat(month.curData[1], "/").concat(_date.Utils.getNumTwoBit(+day.day));
};
var isStart = function isStart(day, days) {
    return _date.Utils.isEqual(days[0], day);
};
var isEnd = function isEnd(day, days) {
    return _date.Utils.isEqual(days[1], day);
};
var isStartAndEnd = function isStartAndEnd(days) {
    return days.length >= 2 && _date.Utils.isEqual(days[0], days[1]);
};
