// @nutui/nutui-react/dist/es/packages/pulltorefresh/pulltorefresh.js
import { _ as _async_to_generator } from "@swc/helpers/_/_async_to_generator";
import { _ as _define_property } from "@swc/helpers/_/_define_property";
import { _ as _instanceof } from "@swc/helpers/_/_instanceof";
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _sliced_to_array } from "@swc/helpers/_/_sliced_to_array";
import { _ as _ts_generator } from "@swc/helpers/_/_ts_generator";
import React, { useEffect, useRef, useState } from "react";
import classNames from "classnames";
import { useDrag } from "@use-gesture/react";
import { animated, useSpring } from "@react-spring/web";
import { Loading, More } from "@nutui/icons-react";
import { useConfig } from "../configprovider";
import { getScrollParent } from "../../utils/get-scroll-parent";
import { rubberbandIfOutOfBounds } from "../../utils/rubberband";
import { sleep } from "../../utils/sleep";
import { passiveSupported } from "../../utils/supports-passive";
import { ComponentDefaults } from "../../utils/typings";
var defaultProps = _object_spread_props(_object_spread({}, ComponentDefaults), {
    type: 'default',
    pullingText: '',
    canReleaseText: '',
    refreshingText: '',
    completeText: '',
    completeDelay: 500,
    disabled: false,
    headHeight: 80,
    threshold: 60,
    onRefresh: function() {}
});
export var PullToRefresh = function(p) {
    var doRefresh = function doRefresh() {
        return _doRefresh.apply(this, arguments);
    };
    var classPrefix = 'nut-pulltorefresh';
    var locale = useConfig().locale;
    var props = _object_spread({}, defaultProps, p, {
        pullingText: p.pullingText || locale.pullToRefresh.pullingText,
        canReleaseText: p.canReleaseText || locale.pullToRefresh.canReleaseText,
        refreshingText: p.refreshingText || locale.pullToRefresh.refreshingText,
        completeText: p.completeText || locale.pullToRefresh.completeText
    });
    var classes = classNames(classPrefix, props.className, "".concat(classPrefix, "-").concat(props.type));
    var headHeight = props.headHeight;
    var threshold = props.threshold;
    var _useState = _sliced_to_array(useState('pulling'), 2), status = _useState[0], setStatus = _useState[1];
    var _useSpring = _sliced_to_array(useSpring(function() {
        return {
            from: {
                height: 0
            },
            config: {
                tension: 300,
                friction: 30,
                clamp: true
            }
        };
    }), 2), springStyles = _useSpring[0], api = _useSpring[1];
    var elementRef = useRef(null);
    var pullingRef = useRef(false);
    useEffect(function() {
        var _elementRef_current;
        (_elementRef_current = elementRef.current) === null || _elementRef_current === void 0 ? void 0 : _elementRef_current.addEventListener('touchmove', function() {});
    }, []);
    function _doRefresh() {
        _doRefresh = _async_to_generator(function() {
            var e;
            return _ts_generator(this, function(_state) {
                switch(_state.label){
                    case 0:
                        api.start({
                            height: headHeight
                        });
                        setStatus('refreshing');
                        _state.label = 1;
                    case 1:
                        _state.trys.push([
                            1,
                            3,
                            ,
                            4
                        ]);
                        return [
                            4,
                            props.onRefresh()
                        ];
                    case 2:
                        _state.sent();
                        setStatus('complete');
                        return [
                            3,
                            4
                        ];
                    case 3:
                        e = _state.sent();
                        api.start({
                            to: /*#__PURE__*/ function() {
                                var _ref = _async_to_generator(function(next) {
                                    return _ts_generator(this, function(_state) {
                                        switch(_state.label){
                                            case 0:
                                                return [
                                                    4,
                                                    next({
                                                        height: 0
                                                    })
                                                ];
                                            case 1:
                                                _state.sent();
                                                setStatus('pulling');
                                                return [
                                                    2
                                                ];
                                        }
                                    });
                                });
                                return function(next) {
                                    return _ref.apply(this, arguments);
                                };
                            }()
                        });
                        throw e;
                    case 4:
                        if (!(props.completeDelay > 0)) return [
                            3,
                            6
                        ];
                        return [
                            4,
                            sleep(props.completeDelay)
                        ];
                    case 5:
                        _state.sent();
                        _state.label = 6;
                    case 6:
                        api.start({
                            to: /*#__PURE__*/ function() {
                                var _ref = _async_to_generator(function(next) {
                                    return _ts_generator(this, function(_state) {
                                        switch(_state.label){
                                            case 0:
                                                return [
                                                    4,
                                                    next({
                                                        height: 0
                                                    })
                                                ];
                                            case 1:
                                                _state.sent();
                                                setStatus('pulling');
                                                return [
                                                    2
                                                ];
                                        }
                                    });
                                });
                                return function(next) {
                                    return _ref.apply(this, arguments);
                                };
                            }()
                        });
                        return [
                            2
                        ];
                }
            });
        });
        return _doRefresh.apply(this, arguments);
    }
    useDrag(function(state) {
        var getScrollTop = function getScrollTop(element) {
            return 'scrollTop' in element ? element.scrollTop : element.scrollY;
        };
        if (status === 'refreshing' || status === 'complete') return;
        var event = state.event;
        // 最后一个事件，检查是否可以刷新或是否是开始状态（第一个状态也是最后一个状态）
        if (state.last) {
            pullingRef.current = false;
            if (status === 'canRelease') {
                doRefresh();
            } else {
                api.start({
                    height: 0
                });
            }
            return;
        }
        // 手指位置
        var _state_movement = _sliced_to_array(state.movement, 2), y = _state_movement[1];
        // 第一个事件，并且手指位置大于0
        if (state.first && y > 0) {
            var target = state.event.target;
            if (!target || !_instanceof(target, Element)) return;
            var scrollParent = getScrollParent(target);
            while(true){
                if (!scrollParent) return;
                var scrollTop = getScrollTop(scrollParent);
                if (scrollTop > 0) {
                    return;
                }
                // 查找到 window 说明到顶了
                if (_instanceof(scrollParent, Window)) {
                    break;
                }
                // 递归查找
                scrollParent = getScrollParent(scrollParent.parentNode);
            }
            pullingRef.current = true;
        }
        if (!pullingRef.current) return;
        if (event.cancelable) {
            event.preventDefault();
        }
        event.stopPropagation();
        var height = Math.max(rubberbandIfOutOfBounds(y, 0, 0, headHeight * 5, 0.5), 0);
        api.start({
            height: height
        });
        setStatus(height > threshold ? 'canRelease' : 'pulling');
    }, {
        pointer: {
            touch: true
        },
        axis: 'y',
        target: elementRef,
        enabled: !props.disabled,
        eventOptions: passiveSupported ? {
            passive: false
        } : false
    });
    var renderIcons = function(status) {
        return /*#__PURE__*/ React.createElement(React.Fragment, null, /*#__PURE__*/ React.createElement("i", {
            className: "".concat(classPrefix, "-head-content-icons")
        }, (status === 'pulling' || status === 'complete') && /*#__PURE__*/ React.createElement(Loading, null), (status === 'canRelease' || status === 'refreshing') && /*#__PURE__*/ React.createElement(More, null)));
    };
    var renderStatusIcon = function() {
        if (props.renderIcon) {
            var _props_renderIcon;
            return (_props_renderIcon = props.renderIcon) === null || _props_renderIcon === void 0 ? void 0 : _props_renderIcon.call(props, status);
        }
        return renderIcons(status);
    };
    var renderStatusText = function() {
        if (props.renderText) {
            var _props_renderText;
            return (_props_renderText = props.renderText) === null || _props_renderText === void 0 ? void 0 : _props_renderText.call(props, status);
        }
        if (status === 'pulling') return props.pullingText;
        if (status === 'canRelease') return props.canReleaseText;
        if (status === 'refreshing') return props.refreshingText;
        if (status === 'complete') return props.completeText;
        return '';
    };
    var _obj, _obj1, _obj2, _obj3, _obj4;
    return /*#__PURE__*/ React.createElement(animated.div, {
        ref: elementRef,
        className: classes,
        style: props.style
    }, /*#__PURE__*/ React.createElement(animated.div, {
        style: springStyles,
        className: classNames((_obj = {}, _define_property(_obj, "".concat(classPrefix, "-head"), true), _define_property(_obj, "".concat(classPrefix, "-primary-head"), props.type === 'primary'), _obj))
    }, /*#__PURE__*/ React.createElement("div", {
        className: classNames((_obj1 = {}, _define_property(_obj1, "".concat(classPrefix, "-head-content"), true), _define_property(_obj1, "".concat(classPrefix, "-primary-head-content"), props.type === 'primary'), _obj1)),
        style: {
            height: headHeight
        }
    }, /*#__PURE__*/ React.createElement("div", {
        className: classNames((_obj2 = {}, _define_property(_obj2, "".concat(classPrefix, "-status-icon"), true), _define_property(_obj2, "".concat(classPrefix, "-primary-status-icon"), props.type === 'primary'), _obj2))
    }, renderStatusIcon()), /*#__PURE__*/ React.createElement("div", {
        className: classNames((_obj3 = {}, _define_property(_obj3, "".concat(classPrefix, "-status-text"), true), _define_property(_obj3, "".concat(classPrefix, "-primary-status-text"), props.type === 'primary'), _obj3))
    }, renderStatusText()))), /*#__PURE__*/ React.createElement("div", {
        className: classNames((_obj4 = {}, _define_property(_obj4, "".concat(classPrefix, "-content"), true), _define_property(_obj4, "".concat(classPrefix, "-primary-content}"), props.type === 'primary'), _obj4))
    }, props.children));
};
PullToRefresh.displayName = 'NutPullToRefresh';
